src/Microsoft.AspNet.Mvc.Razor/Razor/PreCompileViews/RazorPreCompiler.cs

```diff
@@ -58,93 +64,126 @@ public virtual void CompileViews([NotNull] IBeforeCompileContext context)
             }
         }
 
-        protected virtual IReadOnlyList<RazorFileInfo> CreateCompilationDescriptors(
-                                                            [NotNull] IBeforeCompileContext context)
+        protected virtual IEnumerable<RazorFileInfo> CreateCompilationDescriptors(
+            [NotNull] IBeforeCompileContext context)
         {
-            var list = new List<RazorFileInfo>();
+            var filesToProcess = new List<RelativeFileInfo>();
+            GetFileInfosRecursive(root: string.Empty, razorFiles: filesToProcess);
 
-            foreach (var info in GetFileInfosRecursive(string.Empty))
-            {
-                var descriptor = ParseView(info, context);
+            var razorFiles = new RazorFileInfo[filesToProcess.Count];
+            var syntaxTrees = new SyntaxTree[filesToProcess.Count];
+            var parallelOptions = new ParallelOptions { MaxDegreeOfParallelism = MaxDegreesOfParallelism };
 
-                if (descriptor != null)
+            Parallel.For(0, filesToProcess.Count, parallelOptions, index =>
+            {
+                var file = filesToProcess[index];
+                var cacheEntry = PreCompilationCache.GetOrSet(file.RelativePath,
+                                                              file,
+                                                              OnCacheMiss);
+                if (cacheEntry != null)
                 {
-                    list.Add(descriptor);
+                    if (cacheEntry.Success)
+                    {
+                        syntaxTrees[index] = cacheEntry.SyntaxTree;
+                        razorFiles[index] = cacheEntry.FileInfo;
+                    }
+                    else
+                    {
+                        lock (context.Diagnostics)
+                        {
+                            foreach (var diagnostic in cacheEntry.Diagnostics)
+                            {
+                                context.Diagnostics.Add(diagnostic);
+                            }
+                        }
+                    }
                 }
-            }
+            });
+
+            context.CSharpCompilation = context.CSharpCompilation
+                                               .AddSyntaxTrees(syntaxTrees.Where(tree => tree != null));
+            return razorFiles.Where(file => file != null);
+        }
 
-            return list;
+        protected IMvcRazorHost GetRazorHost()
+        {
+            return _serviceProvider.GetRequiredService<IMvcRazorHost>();
         }
 
-        private IEnumerable<RelativeFileInfo> GetFileInfosRecursive(string currentPath)
+        private PrecompilationCacheEntry OnCacheMiss(ICacheSetContext cacheSetContext)
         {
-            var path = currentPath;
```

Comment:
What about compilation options? Is the object thrown away when project.json changes?